### DISCUSSION TITLE
ScratchLab: new block colors

### DISCUSSION THREAD

**Discussion by WorldLanguages** [ScratchAddons/ScratchAddons#5192] [Discussion]
See https://lab.scratch.mit.edu/contrast-blocks/

![image](https://user-images.githubusercontent.com/17484114/195947414-32a6c73e-636f-4e37-ab07-a96994ef3679.png)

> We believe that Scratch should be accessible to everyone, and right now, the colors of our blocks stop that from happening. We know that many of you know and love our current block colors. We hope you can understand that while change can be hard, it's important that everyone can enjoy using Scratch. We appreciate you sharing your thoughts with us in the feedback survey.

Judging by the above paragraph which I took from that same page, Scratch will change the default block colors.
If Scratch doesn't provide an official way to return to the previous block colors, SA will play a big role here. In that case, these should be our goals:
1. Provide an easy way to return to the classic block colors, text color, icon colors, etc. A single click should be enough, even if many addons need to be enabled/settingschanged in order to restore the classic block style (this is related to the concept of a "superpreset").
2. It's likely Scratch users won't like the change at first. We need to try our best to make sure the typical conversation around the change doesn't spin towards "this change is bad", because some kids might not really understand the concept of accessibility and why it's important. When advertising our "restore old blocks" toggle, we need to explain that changing the default is a reasonable decision by the Scratch Team. We could, however, say that an opt-out could have potentially been added, so that's why would be adding it extra-officially as part of the extension.

**Comment by WorldLanguages** [comment]
Another relevant paragraph:

> We hope to introduce these new blocks in the summer of 2023, but we don’t have a launch date yet. We’ll update the Scratch community when we know more.

**Reply by WorldLanguages** [reply]
Summer of 2023 would be around July 2023, I guess.

**Comment by cobaltt7** [comment]
> 1. Provide an easy way to return to the classic block colors, text color, icon colors, etc. A single click should be enough, even if many addons need to be enabled/settingschanged in order to restore the classic block style (this is related to the concept of a "superpreset").

#1934

**Comment by DNin01** [comment]
<sup>I was just writing up an issue about this</sup>
I'm in favor of Scratch's changes (I can read the current blocks fine myself, but I know Scratch really needs the accessibility upgrade), but I know some will prefer the old colors or, more understandably, **dislike them in combination with a dark editor background** (I would). Here's my take:

After these new colors are added to Scratch, the Customizable block colors addon could get a few new preset customizations for the block color design:

- (Default) High contrast light (the new colors - the lighter block colors and black text)
- High contrast dark (think of it like the new colors but inverted, like the old colors but darker with white text - good together with the dark editor theme)

And of course, the other existing presets.

I also noticed Scratch's new colors had some more subtle changes: block outlines are darker, and text is black (like, `#000000` black), whereas this addon's black setting makes it dark gray. I'd assume we'll need to take those details into account, even if the user can't customize them.

**Reply by DNin01** [reply]
I customized the addon to match the new colors as close as possible and changed the code to make black text actually black. Here's what it looks like (on the left) compared to what they look like on Scratch Lab (on the right).
![A picture of my edit versus the official high contrast block colors](https://user-images.githubusercontent.com/106490990/197311269-f78da76e-ccd5-4397-af24-ef7dd847014b.png)
We should make sure these are identical, since users would assume the addon's default preset is unchanged.

**Reply by penguinmoose** [reply]
There should be an option in the addon to make the icons and symbols on the blocks (like the turn left/right arrows) the same as the block text colors.

**Reply by mxmou** [reply]
> There should be an option in the addon to make the icons and symbols on the blocks (like the turn left/right arrows) the same as the block text colors.

It doesn't even have to be an option because the current behavior doesn't make much sense.

**Reply by penguinmoose** [reply]
Then the coloring should be consistent. There doesn't even need to be an option.

**Comment by mxmou** [comment]
We could also implement the new colors as a preset before they're added to Scratch and become the default.

**Reply by WorldLanguages** [reply]
I think we should allow that possibility. As of now, there's no way to use SA to have black icons, and #000000 block text, AFAIK.

**Reply by mxmou** [reply]
I think I'll open a PR to change the black text option.

**Reply by DNin01** [reply]
And I have two new presets ready, one that is the same as the new block colors and one that I made with dark versions of those high contrast colors. I can open a PR now and you can merge it when you're ready or wait until later to open it. We'll still need to update other details, as I mentioned: https://github.com/ScratchAddons/ScratchAddons/discussions/5192#discussioncomment-3938133

**Reply by DNin01** [reply]
@WorldLanguages It sounds like you want to have a preset to recreate the new block colors ahead of time - would you like if I make a PR to add a new preset (and a dark variation of it) now so users have the option to use the updated block colors before the block colors update? Or are we waiting until close to when Scratch updates the block colors? Have a plan?

**Reply by WorldLanguages** [reply]
@DNin01 We could add a preset within #5257 I think

**Reply by penguinmoose** [reply]
Why would the black text become the default before Scratch even implements it? Or do you mean the preset will only become default once Scratch applies the high-contrast styling?

**Reply by WorldLanguages** [reply]
@penguinmoose Your question is confusing. But I think you misunderstood what mxmou meant.
These new block colors will be Scratch's default. We don't know yet if Scratch themselves will offer a way to return to the old colors.

**Reply by mxmou** [reply]
@penguinmoose The preset will become the default when Scratch implements it, but we might add it earlier as an option.

**Reply by penguinmoose** [reply]
@mxmou Yes, we should add it earlier as an option - after all, some people may want it before Scratch officially releases it (if ever).

**Reply by penguinmoose** [reply]
@WorldLanguages That's true, if Scratch offers a way to return to the old colors, do you think we still should have this preset?

**Reply by BroJac5246** [reply]
> @WorldLanguages That's true, if Scratch offers a way to return to the old colors, do you think we still should have this preset?

Yes, since some people will want to modify Scratch's defaults. Additionally, that's an _if_, which we can't count on.

**Reply by penguinmoose** [reply]
@BroJac5246 Good point, but we will have the current preset available either way.

Also, I contacted the ST and asked them to consider adding a way to return to the old colors, but it doesn't really matter as far as we are concerned.

**Reply by WorldLanguages** [reply]
They probably have their own reasons not to add an option, and that's understandable.
On the other hand, we have no choice but to add the option, if we're being honest. It's the opposite situation.

**Reply by penguinmoose** [reply]
That’s true. But if they do offer an option to turn high contrast blocks off, there will be two ways to turn it off: through the website and through Scratch Addons. I don’t think we should have duplicates, so should we hide the option so it’s consistent?

**Reply by DNin01** [reply]
> That’s true. But if they do offer an option to turn high contrast blocks off, there will be two ways to turn it off: through the website and through Scratch Addons. I don’t think we should have duplicates, so should we hide the option so it’s consistent?

If we don't offer a way to use the original colors in SA, then it will make it harder for users to make modifications of the original colors.

**Reply by penguinmoose** [reply]
> If we don't offer a way to use the original colors in SA, then it will make it harder for users to make modifications of the original colors.

If we keep both options (if they exist), what if the website option is set to regular mode, and SA is set to high contrast preset (or vise versa)? Does SA override the website setting - if one is offered?

**Reply by BroJac5246** [reply]
> > If we don't offer a way to use the original colors in SA, then it will make it harder for users to make modifications of the original colors.
> 
> If we keep both options (if they exist), what if the website option is set to regular mode, and SA is set to high contrast preset (or vise versa)? Does SA override the website setting - if one is offered?

Yes, SA would override whatever the website is using– that's the point of the addon. If it didn't then the addon wouldn't do anything, even if the block design that the user wants to use isn't a preset.

**Comment by DNin01** [comment]
Could we create a new branch at some point so we can work on these changes ahead of time?
<sub>(I've got my new presets just laying around waiting for a chance...)</sub>

**Reply by WorldLanguages** [reply]
What do you mean? A new branch for what?

**Reply by DNin01** [reply]
Collecting changes to the customizable block colors addon (e.g. if it takes multiple contributors to update the addon to work well with Scratch's upcoming block colors, then we could keep those changes on a separate branch until we're ready to release them).

Though, it might be easier to start the changes during a Scratch Addons update around the time the new block colors come out so that we don't have to do any branching.

**Reply by WorldLanguages** [reply]
I'd suggest having 1 open PR for that, and send pull requests to the correct branch on the SA fork from the PR author.

**Reply by DNin01** [reply]
If I'm getting this right, then I could make a branch and a PR that other contributors could then make PRs to add their own changes to?

And it would work like this:
- I make a `change-block-colors` branch on my fork
- I send a PR
- Some other contributors send PRs to change code on my branch
- Those PRs are merged
- When my PR is merged, everyone else's changes come along

**Comment by cobaltt7** [comment]
You can create as many personal branches on your own fork as you want, that’s how GitHub works…

**Reply by DNin01** [reply]
I meant a branch on the origin repo, ScratchAddons/ScratchAddons.

**Comment by cobaltt7** [comment]
Branches on the origin repo are discouraged even among the organization members…why can’t you use a fork?

**Reply by DNin01** [reply]
Because I won't be the only one working on it. I was actually thinking about a few of us developing these changes overtime on a branch and then pushing them to the main branch when they're ready.

Though, it might be easier to just start working on everything for a certain Scratch Addons update around the time the new block colors come out. Then we wouldn't have to do any branching on the origin repo.

**Comment by DNin01** [comment]
This is interesting... it seems like they're working on code to allow changing the "theme": https://github.com/LLK/scratch-gui/pull/8753 so users can decide whether they want high contrast block colors on or not.

**Reply by BroJac5246** [reply]
Also interesting is that 'dark-mode'
